src/main/java/net/fabricmc/loader/impl/FabricLoaderImpl.java

```diff
@@ -177,36 +183,66 @@ private void setup() throws ModResolutionException {
 		discoverer.addCandidateFinder(new ClasspathModCandidateFinder());
 		discoverer.addCandidateFinder(new DirectoryModCandidateFinder(gameDir.resolve("mods"), isDevelopmentEnvironment()));
 
-		Map<String, ModCandidateSet> mods = discoverer.discoverMods(this);
-		Map<String, ModCandidate> candidateMap = new ModResolver().resolve(mods);
+		List<ModCandidate> mods = ModResolver.resolve(discoverer.discoverMods(this));
 
-		String modListText = candidateMap.values().stream()
-				.sorted(Comparator.comparing(candidate -> candidate.getInfo().getId()))
-				.map(candidate -> String.format("\t- %s@%s", candidate.getInfo().getId(), candidate.getInfo().getVersion().getFriendlyString()))
+		String modListText = mods.stream()
+				.map(candidate -> String.format("\t- %s %s", candidate.getId(), candidate.getVersion().getFriendlyString()))
 				.collect(Collectors.joining("\n"));
 
-		int count = candidateMap.values().size();
+		int count = mods.size();
 		Log.info(LogCategory.GENERAL, "Loading %d mod%s:%n%s", count, count != 1 ? "s" : "", modListText);
 
 		if (DependencyOverrides.INSTANCE.getDependencyOverrides().size() > 0) {
-			Log.info(LogCategory.GENERAL, "Dependencies overridden for \"%s\"", String.join(", ", DependencyOverrides.INSTANCE.getDependencyOverrides().keySet()));
+			Log.info(LogCategory.GENERAL, "Dependencies overridden for \"%s\"", String.join(", ", DependencyOverrides.INSTANCE.getDependencyOverrides()));
+		}
+
+		Path cacheDir = gameDir.resolve(CACHE_DIR_NAME);
+		Path outputdir = cacheDir.resolve(PROCESSED_MODS_DIR_NAME);
+
+		// runtime mod remapping
+
+		if (isDevelopmentEnvironment()) {
+			if (System.getProperty(SystemProperties.REMAP_CLASSPATH_FILE) == null) {
+				Log.warn(LogCategory.MOD_REMAP, "Runtime mod remapping disabled due to no fabric.remapClasspathFile being specified. You may need to update loom.");
+			} else {
+				RuntimeModRemapper.remap(mods, cacheDir.resolve(TMP_DIR_NAME), outputdir);
+			}
 		}
 
-		boolean runtimeModRemapping = isDevelopmentEnvironment();
+		// shuffle mods in-dev to reduce the risk of false order reliance, apply late load requests
 
-		if (runtimeModRemapping && System.getProperty(SystemProperties.REMAP_CLASSPATH_FILE) == null) {
-			Log.warn(LogCategory.MOD_REMAP, "Runtime mod remapping disabled due to no fabric.remapClasspathFile being specified. You may need to update loom.");
-			runtimeModRemapping = false;
+		if (isDevelopmentEnvironment()) {
```

Comment:
Not sure if we want a proptery to turn this off, I can possibly see it becoming annoying when trying to debug something.

Reply:
added a system property to disable it